doh-proxy-rust: init at 0.3.8

stephank opened the pull request:
##### Motivation for this change

Using this on my own NixOS server, thought I'd contribute back. Note that this is my first contribution!

Hopefully, `doh-proxy-rust` is the right naming for this package. It conflicts with [facebookexperimental/doh-proxy](https://github.com/facebookexperimental/doh-proxy). (already packaged)

I've added myself as maintainer, which I'm happy to do, but am also unsure how to best approach. For example, I don't actively monitor upstream for releases, but am curious if this is automated somehow. But I'm at least available to review issues/PRs whenever tagged, if that helps.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

fzakaria commented on maintainers/maintainer-list.nix:
```diff
@@ -7598,6 +7598,12 @@
     githubId = 1699155;
     name = "Steve Elliott";
   };
+  stephank = {
+    email = "nix@stephank.nl";
+    github = "stephank";
+    githubId = 89950;
+    name = "Stéphan Kochen";
+  };
```
I've seen reviewers preferring this to be a separate commit.
(the commit tag line has a format when adding new maintainers)

fzakaria reviewed (commented):
the derivation looks good to me; but i don't have familiarity enough on NixOS to comment on the module.
I would rebase your commit though so it's:
1. add maintainer
2. init doh-proxy

Looks like nixpkgs is already getting into namespace issues :(

stephank committed:
maintainers: add stephank

stephank committed:
doh-proxy-rust: init at 0.3.8

stephank commented:
Thanks for the review! I rebased on master, redid the commits, and also updated `Cargo.lock` for minor dep upgrades. (The original project doesn't have a lockfile.)

SuperSandro2000 commented on pkgs/servers/dns/doh-proxy-rust/default.nix:
```diff
@@ -0,0 +1,28 @@
+    description = "Fast, mature, secure DoH server proxy written in Rust";
+    license = with licenses; [ mit ];
+    maintainers = with maintainers; [ stephank ];
+    platforms = platforms.all;
```
```suggestion
```

stephank replied:
I rebased on master with this change.

SuperSandro2000 commented:
@ofborg eval

stephank commented:
I rebased this on master and upgraded to 0.3.8, but the test still needs some work. Will need a bit of time to look into this.

stephank commented:
Turns out, the original test was also flawed, but miraculously succeeded. This is now fixed.

SuperSandro2000 commented on pkgs/servers/dns/doh-proxy-rust/cargo-lock.patch:
```diff
@@ -0,0 +1,838 @@
+diff --git a/Cargo.lock b/Cargo.lock
```
It would be great if you could try to upstream this.

stephank replied:
Unfortunately, upstream has rejected this in the past: https://github.com/jedisct1/doh-server/issues/31

SuperSandro2000 commented on nixos/tests/doh-proxy-rust.nix:
```diff
@@ -0,0 +1,43 @@
+import ./make-test-python.nix ({ pkgs, ... }: {
+  name = "doh-proxy-rust";
+  meta = with pkgs.stdenv.lib.maintainers; {
```
```suggestion
  meta = with lib.maintainers; {
```

SuperSandro2000 commented on nixos/tests/doh-proxy-rust.nix:
```diff
@@ -0,0 +1,43 @@
+import ./make-test-python.nix ({ pkgs, ... }: {
```
```suggestion
import ./make-test-python.nix ({ lib, pkgs, ... }: {
```

SuperSandro2000 reviewed (changes requested):
@aanderse could you take a 2nd look at the module?

aanderse commented on nixos/modules/services/networking/doh-proxy-rust.nix:
```diff
@@ -0,0 +1,69 @@
+
+  config = mkIf cfg.enable {
+
+    users.users.doh-proxy-rust = {
```
Maybe you just want to use `DynamicUser` instead.

stephank replied:
This doesn't work together with `confinement`, but per the other comment, I've disabled `confinement` for now.

aanderse replied:
Alright.

aanderse commented on nixos/modules/services/networking/doh-proxy-rust.nix:
```diff
@@ -0,0 +1,69 @@
+      description = "doh-proxy-rust";
+      after = [ "network.target" "nss-lookup.target" ];
+      wantedBy = [ "multi-user.target" ];
+      confinement = {
```
While I'm sure @aszlig would love to see this... we don't actually use `confinement` anywhere else in our module system.

@flokli @andir @arianvp @Mic92 what do you all think about using `confinement` in upstream NixOS modules?

aanderse commented on nixos/modules/services/networking/doh-proxy-rust.nix:
```diff
@@ -0,0 +1,69 @@
+      default = [];
+      example = literalExample [ "--server-address=9.9.9.9:53" ];
+      description = ''
+        A list of command-line flags to pass to doh-proxy.
```
Including a link to https://github.com/jedisct1/doh-server#usage would be useful.

stephank replied:
I hope I'm doing this correctly. I've been looking through the docs, but can't figure out how the options index on nixos.org is rendered, and how I would preview this locally. Could you point me in the right direction?

aanderse replied:
Yes that is correct. You can render the manual and take a look with this: `~/nixpkgs (master)> nix-build nixos/release.nix -A manual.x86_64-linux`.

aanderse reviewed (commented):
A nice and simple module 👍 Just a few points for discussion.

aanderse commented:
Looks good @stephank :+1:
